src/coreclr/src/jit/compiler.cpp

```diff
@@ -960,6 +960,16 @@ var_types Compiler::getReturnTypeForStruct(CORINFO_CLASS_HANDLE clsHnd,
     // The largest "primitive type" is MAX_PASS_SINGLEREG_BYTES
     // so we can skip calling getPrimitiveTypeForStruct when we
     // have a struct that is larger than that.
+
+#ifdef _TARGET_AMD64_
+    if ((impNormStructType(clsHnd) == TYP_SIMD16))
```

Comment:
This should probably be handled in `getPrimitiveTypeForStruct` but requires some work there, because IIRC windows and linux have different ABI requirements for args (and this same method is used for both).

Reply:
Maybe I'm missing something but won't this simple `TYP_SIMD16` check result in `Vector4` being returned in `xmm0` instead of the ret buffer?

Reply:
Yes, good point @mikedn, but I think that requires verification - my recollection is that the `float4` struct in Visual C++ (which `Vector4` was designed to be compatible with) is passed as a "regular" struct of 16 bytes, but I'm not entirely certain of that.

Reply:
I'm not sure if you have a particular `float4` struct in mind but VC++ does return a normal 4 float struct via the return buffer:
https://gcc.godbolt.org/z/nSrf85

It returns it in XMM0 only if the vectorcall calling convention is used:
https://gcc.godbolt.org/z/HiKuxD

Reply:
I believe, under previous discussions, we had said `Vector4` was to be a user-defined struct of `4 floats`, which would make it at best an HFA (being passed in 4 registers under `__vectorcall` and 2 registers under `SysV`)

Reply:
This was also needed for back-compat as `Vector2/3/4` aren't generic and so have been usable in P/Invoke declarations since first implemented (and so the expecation is it is treated as a user-defined struct).

Reply:
> It returns it in XMM0 only if the vectorcall calling convention is used:

It's worth noting that its passed/returned in 4 registers; not just xmm0:
https://gcc.godbolt.org/z/YQ_Hgd

Reply:
Right, I keep forgetting that while such structs are given special status they're not really treated as a single vector. That will only make implementing vectorcall in the JIT more complicated...

Reply:
True, though we've already got HFAs for arm & arm64, so some of the hard work is already done. Maybe we can hope for / actually achieve some cleanup of ABI handling that can actually leverage that.